### Acesso arquivos séries históricas B3

Link para baixar os arquivos: http://www.b3.com.br/pt_br/market-data-e-indices/servicos-de-dados/market-data/historico/mercado-a-vista/series-historicas/  

Link com informações para a leitura dos arquivos: http://www.b3.com.br/pt_br/market-data-e-indices/servicos-de-dados/market-data/historico/mercado-a-vista/cotacoes-historicas/  

A B3 disponbiliza suas séries históricas na página acima, sendo que é necessário fazer cadastro, e informar um *captcha* para baixar os arquivos anuais, que estão zipados.  
Os passos abaixo consideram que os arquivos foram baixados e descompactados manualmente, e o arquivo da série referente ao ano está no mesmo diretório do *notebook*.

In [1]:
import pandas as pd

A *string* abaixo é um exemplo de um registro do arquivo de séries:

    012015042222MYPK11      010IOCHP-MAXIONBNS ORD NM   R$  000000000003000000000000300000000000030000000000003000000000000300000000000030000000000025000001000000000000015800000000000000474000000000000000009999123100000010000000000000BRMYPKN01OR1100

Este é um tipo de arquivo chamado de "comprimento fixo". O formato texto de comprimento fixo destina um número específico de caracteres por campo, não sendo utilizados separadores entre campos. Todos os registros tem o mesmo número de caracteres ainda que os valores dos registros caibam em menos caracteres.

Para ler este tipo de arquivo utiliza-se a função [pandas.read_fwf](https://pandas.pydata.org/docs/reference/api/pandas.read_fwf.html)  
Seguindo http://jonathansoma.com/lede/foundations-2017/pandas/opening-fixed-width-files/  
1) Criar uma lista com os comprimentos dos campos:

In [2]:
comprimentos = [
2,  # TIPREG - TIPO DE REGISTRO - FIXO “01”
8,  # DATA-P - DATA DO PREGÃO - FORMATO “AAAAMMDD” 
2,  # CODBDI - CÓDIGO BDI - UTILIZADO PARA CLASSIFICAR OS PAPÉIS NA EMISSÃO DO BOLETIM DIÁRIO DE INFORMAÇÕES
12, # CODNEG - CÓDIGO DE NEGOCIAÇÃO DO PAPEL 
3,  # TPMERC - TIPO DE MERCADO - CÓD. DO MERCADO EM QUE O PAPEL ESTÁ CADASTRADO
12, # NOMRES - NOME RESUMIDO DA EMPRESA EMISSORA DO PAPEL 
10, # ESPECI - ESPECIFICAÇÃO DO PAPEL 
3,  # PRAZOT - PRAZO EM DIAS DO MERCADO A TERMO 
4,  # MODREF - MOEDA DE REFERÊNCIA - USADA NA DATA DO PREGÃO 
13, # PREABE - PREÇO DE ABERTURA PAPELMERCADO PREGÃO DO NO 
13, # PREMAX - PREÇO MÁXIMO PAPELMERCADO PREGÃO DO NO 
13, # PREMIN - PREÇO MÍNIMO DO PAPEL- MERCADO NO PREGÃO 
13, # PREMED - PREÇO MÉDIO DO PAPEL- MERCADO NO PREGÃO 
13, # PREULT - PREÇO DO ÚLTIMO NEGÓCIO DO PAPEL-MERCADO NO PREGÃO 
13, # PREOFC - PREÇO DA MELHOR OFERTA DE COMPRA DO PAPELMERCADO 
13, # PREOFV - PREÇO DA MELHOR OFERTA DE VENDA DO PAPELMERCADO 
5,  # TOTNEG - NÚMERO DE NEGÓCIOS EFETUADOS COM O PAPEL- MERCADO NO PREGÃO 
18, # QUATOT - QUANTIDADE TOTAL DE TÍTULOS NEGOCIADOS NESTE PAPEL- MERCADO 
18, # VOLTOT - VOLUME TOTAL DE TÍTULOS NEGOCIADOS NESTE PAPEL- MERCADO 
13, # PREEXE - PREÇO DE EXERCÍCIO PARA O MERCADO DE OPÇÕES OU VALOR DO CONTRATO PARA O MERCADO DE TERMO SECUNDÁRIO 
1,  # INDOPC - INDICADOR DE CORREÇÃO DE PREÇOS DE EXERCÍCIOS OU VALORES DE CONTRATO PARA OS MERCADOS DE OPÇÕES OU TERMO SECUNDÁRIO 
8,  # DATVEN - INDICADOR DE CORREÇÃO DE PREÇOS DE EXERCÍCIOS OU VALORES DE CONTRATO PARA OS MERCADOS DE OPÇÕES OU TERMO SECUNDÁRIO - FORMATO “AAAAMMDD”  
7,  # FATCOT - FATOR DE COTAÇÃO DO PAPEL - ‘1’ = COTAÇÃO UNITÁRIA ‘1000’ = COTAÇÃO POR LOTE DE MIL AÇÕES 
13, # PTOEXE - PREÇO DE EXERCÍCIO EM PONTOS PARA OPÇÕES REFERENCIADAS EM DÓLAR OU VALOR DE CONTRATO EM PONTOS PARA TERMO SECUNDÁRIO (ver planilha)
12, # CODISI - CÓDIGO DO PAPEL NO SISTEMA ISIN OU CÓDIGO INTERNO DO PAPEL (ver planilha) 
3,  # DISMES - NÚMERO DE DISTRIBUIÇÃO DO PAPEL (ver planilha) 
]
#print(comprimentos)

2) Criando uma lista com o nome das colunas.  
   A primeira linha do arquivo não é o nome das colunas (ver o documento da B3 referido acima) então precisamos nomear elas após a leitura do arquivo. Estou usando o código abreviado da referência.

In [3]:
colunas = ['TIPREG', 'DATA-P', 'CODBDI', 'CODNEG', 'TPMERC', 'NOMRES', 'ESPECI', 'PRAZOT', 'MODREF', 'PREABE', 'PREMAX', 'PREMIN', 'PREMED', 'PREULT', 'PREOFC', 'PREOFV', 'TOTNEG', 'QUATOT', 'VOLTOT', 'PREEXE', 'INDOPC', 'DATVEN', 'FATCOT', 'PTOEXE', 'CODISI', 'DISMES']
print(colunas)

['TIPREG', 'DATA-P', 'CODBDI', 'CODNEG', 'TPMERC', 'NOMRES', 'ESPECI', 'PRAZOT', 'MODREF', 'PREABE', 'PREMAX', 'PREMIN', 'PREMED', 'PREULT', 'PREOFC', 'PREOFV', 'TOTNEG', 'QUATOT', 'VOLTOT', 'PREEXE', 'INDOPC', 'DATVEN', 'FATCOT', 'PTOEXE', 'CODISI', 'DISMES']


3) Lendo o arquivo **(lembrar que os arquivos são grandes; demora e o dataframe criado ocupa bastante memória)**:

In [4]:
teste = pd.read_fwf("COTAHIST_A2015.TXT", widths=comprimentos)

4) A primeira linha não tem relação direta com as colunas:

In [5]:
teste.head()

00  COTAHIST    .2 015BOVESPA 2  015          1230  Unnamed: 6  Unnamed: 7  \
0   1  20150102   2.0       AAPL34   10         APPLE         DRN         NaN   
1   1  20150102   2.0        ABCB4   10    ABC BRASIL  PN  EJS N2         NaN   
2   1  20150102  96.0       ABCB4F   20    ABC BRASIL  PN  EJS N2         NaN   
3   1  20150102  12.0       ABCP11   10  FII ABC IMOB      CI  ER         NaN   
4   1  20150102   2.0        ABEV3   10     AMBEV S/A      ON  EJ         NaN   

  Unnamed: 8  Unnamed: 9  ...  Unnamed: 16  Unnamed: 17   Unnamed: 18  \
0         R$      2965.0  ...          2.0      32300.0  9.564350e+07   
1         R$      1285.0  ...        531.0     117100.0  1.469337e+08   
2         R$      1283.0  ...          5.0        200.0  2.641200e+05   
3         R$       972.0  ...          9.0        490.0  4.781400e+05   
4         R$      1614.0  ...      21543.0   19110800.0  3.073420e+10   

   Unnamed: 19  Unnamed: 20  Unnamed: 21  Unnamed: 22  Unnamed: 23  \
0          0.0          0.0   99991231.0          1.0          0.0   
1          0.0          0.0   99991231.0          1.0          0.0   
2          0.0          0.0   99991231.0          1.0          0.0   
3          0.0          0.0   99991231.0          1.0          0.0   
4          0.0          0.0   99991231.0          1.0          0.0   

    Unnamed: 24  Unnamed: 25  
0  BRAAPLBDR004        111.0  
1  BRABCBACNPR4        123.0  
2  BRABCBACNPR4        123.0  
3  BRABCPCTF000        231.0  
4  BRABEVACNOR1        106.0  

[5 rows x 26 columns]

5) A última linha não é um registro e sim uma "linha de fim" do formato da B3:

In [6]:
teste.tail()

00  COTAHIST    .2  015BOVESPA 2  015          1230 Unnamed: 6  \
414173   1  20151230  14.0        XPTD11   10   FIA TOP DIV         CI   
414174   1  20151230   2.0        XRXB34   10    XEROX CORP        DRN   
414175   1  20151230  96.0       XRXB34F   20    XEROX CORP        DRN   
414176   1  20151230  12.0        XTED11   10  FII TRXE COR         CI   
414177  99  COTAHIST   0.2  015BOVESPA 2   15  123000000414        179   

        Unnamed: 7 Unnamed: 8  Unnamed: 9  ...  Unnamed: 16  Unnamed: 17  \
414173         NaN         R$      7709.0  ...        849.0       3130.0   
414174         NaN         R$      4264.0  ...          1.0        300.0   
414175         NaN         R$      4264.0  ...          1.0          0.0   
414176         NaN         R$      4000.0  ...         36.0       2300.0   
414177         NaN        NaN         NaN  ...          NaN          NaN   

        Unnamed: 18  Unnamed: 19  Unnamed: 20  Unnamed: 21  Unnamed: 22  \
414173   24441120.0          0.0          0.0   99991231.0          1.0   
414174    1279200.0          0.0          0.0   99991231.0          1.0   
414175      17050.0          0.0          0.0   99991231.0          1.0   
414176    9127900.0          0.0          0.0   99991231.0          1.0   
414177          NaN          NaN          NaN          NaN          NaN   

        Unnamed: 23   Unnamed: 24  Unnamed: 25  
414173          0.0  BRXPTDCTF002        165.0  
414174          0.0  BRXRXBBDR005        109.0  
414175          0.0  BRXRXBBDR005        109.0  
414176          0.0  BRXTEDCTF006        135.0  
414177          NaN           NaN          NaN  

[5 rows x 26 columns]

5a) Apagando a última linha:

In [7]:
teste.drop(teste.tail(1).index,inplace=True)
teste.tail()

00  COTAHIST    .2 015BOVESPA 2  015          1230 Unnamed: 6  \
414172   1  20151230  14.0       XPOM11   10  FIP XP OMEGA         CI   
414173   1  20151230  14.0       XPTD11   10   FIA TOP DIV         CI   
414174   1  20151230   2.0       XRXB34   10    XEROX CORP        DRN   
414175   1  20151230  96.0      XRXB34F   20    XEROX CORP        DRN   
414176   1  20151230  12.0       XTED11   10  FII TRXE COR         CI   

        Unnamed: 7 Unnamed: 8  Unnamed: 9  ...  Unnamed: 16  Unnamed: 17  \
414172         NaN         R$      7400.0  ...          8.0       1200.0   
414173         NaN         R$      7709.0  ...        849.0       3130.0   
414174         NaN         R$      4264.0  ...          1.0        300.0   
414175         NaN         R$      4264.0  ...          1.0          0.0   
414176         NaN         R$      4000.0  ...         36.0       2300.0   

        Unnamed: 18  Unnamed: 19  Unnamed: 20  Unnamed: 21  Unnamed: 22  \
414172    8924000.0          0.0          0.0   99991231.0          1.0   
414173   24441120.0          0.0          0.0   99991231.0          1.0   
414174    1279200.0          0.0          0.0   99991231.0          1.0   
414175      17050.0          0.0          0.0   99991231.0          1.0   
414176    9127900.0          0.0          0.0   99991231.0          1.0   

        Unnamed: 23   Unnamed: 24  Unnamed: 25  
414172          0.0  BRXPOMCTF004        105.0  
414173          0.0  BRXPTDCTF002        165.0  
414174          0.0  BRXRXBBDR005        109.0  
414175          0.0  BRXRXBBDR005        109.0  
414176          0.0  BRXTEDCTF006        135.0  

[5 rows x 26 columns]

6) Renomeando as colunas:

In [8]:
teste.columns = colunas
teste.head()

TIPREG    DATA-P  CODBDI  CODNEG  TPMERC        NOMRES      ESPECI  PRAZOT  \
0       1  20150102     2.0  AAPL34      10         APPLE         DRN     NaN   
1       1  20150102     2.0   ABCB4      10    ABC BRASIL  PN  EJS N2     NaN   
2       1  20150102    96.0  ABCB4F      20    ABC BRASIL  PN  EJS N2     NaN   
3       1  20150102    12.0  ABCP11      10  FII ABC IMOB      CI  ER     NaN   
4       1  20150102     2.0   ABEV3      10     AMBEV S/A      ON  EJ     NaN   

  MODREF  PREABE  ...   TOTNEG      QUATOT        VOLTOT  PREEXE  INDOPC  \
0     R$  2965.0  ...      2.0     32300.0  9.564350e+07     0.0     0.0   
1     R$  1285.0  ...    531.0    117100.0  1.469337e+08     0.0     0.0   
2     R$  1283.0  ...      5.0       200.0  2.641200e+05     0.0     0.0   
3     R$   972.0  ...      9.0       490.0  4.781400e+05     0.0     0.0   
4     R$  1614.0  ...  21543.0  19110800.0  3.073420e+10     0.0     0.0   

       DATVEN  FATCOT  PTOEXE        CODISI  DISMES  
0  99991231.0     1.0     0.0  BRAAPLBDR004   111.0  
1  99991231.0     1.0     0.0  BRABCBACNPR4   123.0  
2  99991231.0     1.0     0.0  BRABCBACNPR4   123.0  
3  99991231.0     1.0     0.0  BRABCPCTF000   231.0  
4  99991231.0     1.0     0.0  BRABEVACNOR1   106.0  

[5 rows x 26 columns]

7) Tratar a data:  
referências  
https://pbpython.com/pandas_dtypes.html  
https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.to_datetime.html  
https://docs.python.org/3/library/datetime.html#strftime-and-strptime-behavior  

In [9]:
teste.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 414177 entries, 0 to 414176
Data columns (total 26 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   TIPREG  414177 non-null  int64  
 1   DATA-P  414177 non-null  object 
 2   CODBDI  414177 non-null  float64
 3   CODNEG  414177 non-null  object 
 4   TPMERC  414177 non-null  int64  
 5   NOMRES  414177 non-null  object 
 6   ESPECI  414177 non-null  object 
 7   PRAZOT  249163 non-null  float64
 8   MODREF  414177 non-null  object 
 9   PREABE  414177 non-null  float64
 10  PREMAX  414177 non-null  float64
 11  PREMIN  414177 non-null  float64
 12  PREMED  414177 non-null  float64
 13  PREULT  414177 non-null  float64
 14  PREOFC  414177 non-null  float64
 15  PREOFV  414177 non-null  float64
 16  TOTNEG  414177 non-null  float64
 17  QUATOT  414177 non-null  float64
 18  VOLTOT  414177 non-null  float64
 19  PREEXE  414177 non-null  float64
 20  INDOPC  414177 non-null  float64
 21  DATVEN  41

In [10]:
teste['DATA-P'] = pd.to_datetime(teste['DATA-P'], format='%Y%m%d')
teste.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 414177 entries, 0 to 414176
Data columns (total 26 columns):
 #   Column  Non-Null Count   Dtype         
---  ------  --------------   -----         
 0   TIPREG  414177 non-null  int64         
 1   DATA-P  414177 non-null  datetime64[ns]
 2   CODBDI  414177 non-null  float64       
 3   CODNEG  414177 non-null  object        
 4   TPMERC  414177 non-null  int64         
 5   NOMRES  414177 non-null  object        
 6   ESPECI  414177 non-null  object        
 7   PRAZOT  249163 non-null  float64       
 8   MODREF  414177 non-null  object        
 9   PREABE  414177 non-null  float64       
 10  PREMAX  414177 non-null  float64       
 11  PREMIN  414177 non-null  float64       
 12  PREMED  414177 non-null  float64       
 13  PREULT  414177 non-null  float64       
 14  PREOFC  414177 non-null  float64       
 15  PREOFV  414177 non-null  float64       
 16  TOTNEG  414177 non-null  float64       
 17  QUATOT  414177 non-null  floa

7a) A coluna DATVEN tem algumas datas com o valor `99991231` que seria a data 31/12/9999. Acredito que seja assim para indicar ativos sem data de vencimento. Só que o Python não consegue tratar esta data, ver abaixo.

In [11]:
teste['DATVEN'].value_counts()

99991231.0    192692
20151221.0     23477
20150921.0     18099
20150420.0     17599
20151019.0     17250
20151116.0     15836
20150518.0     15291
20150720.0     15144
20150615.0     14557
20150817.0     14537
20160118.0     14332
20150316.0     13961
20150209.0      8243
20160215.0      7151
20150119.0      3948
20160321.0      2859
20151216.0      2671
20160418.0      2145
20151014.0      1722
20160217.0      1652
20150812.0      1251
20150617.0      1231
20160413.0      1082
20150415.0       849
20161219.0       630
20160620.0       585
20160516.0       580
20160919.0       557
20161017.0       545
20160718.0       497
20150218.0       456
20160815.0       406
20161121.0       333
20150916.0       311
20151118.0       293
20170116.0       258
20160817.0       214
20150715.0       174
20160113.0       156
20160615.0       142
20170320.0       105
20161214.0       101
20170220.0        86
20150513.0        73
20170619.0        40
20170614.0        21
20171121.0        10
20161013.0   

https://pandas.pydata.org/pandas-docs/stable/user_guide/timeseries.html#timestamp-limitations  
**Limites para timestamp:**  
Since pandas represents timestamps in nanosecond resolution, the time span that can be represented using a 64-bit integer is limited to approximately 584 years:  

    In [92]: pd.Timestamp.min
    Out[92]: Timestamp('1677-09-21 00:12:43.145225')

    In [93]: pd.Timestamp.max
    Out[93]: Timestamp('2262-04-11 23:47:16.854775807')

Aqui (https://pandas.pydata.org/pandas-docs/stable/user_guide/timeseries.html#timeseries-oob) sugere maneiras de tratar "out-of-bound" mas não acho necessário neste caso.  
**Para manter o dado no mesmo formato, acredito que subsituir "99991231" por um valor próximo a "Timestamp.max" é suficiente, indicando a substituição**.  

Será preciso verificar durante a utilização dos valores. 

In [12]:
teste.loc[(teste['DATVEN'] == 99991231),'DATVEN'] = 22611231
teste['DATVEN'].value_counts()

22611231.0    192692
20151221.0     23477
20150921.0     18099
20150420.0     17599
20151019.0     17250
20151116.0     15836
20150518.0     15291
20150720.0     15144
20150615.0     14557
20150817.0     14537
20160118.0     14332
20150316.0     13961
20150209.0      8243
20160215.0      7151
20150119.0      3948
20160321.0      2859
20151216.0      2671
20160418.0      2145
20151014.0      1722
20160217.0      1652
20150812.0      1251
20150617.0      1231
20160413.0      1082
20150415.0       849
20161219.0       630
20160620.0       585
20160516.0       580
20160919.0       557
20161017.0       545
20160718.0       497
20150218.0       456
20160815.0       406
20161121.0       333
20150916.0       311
20151118.0       293
20170116.0       258
20160817.0       214
20150715.0       174
20160113.0       156
20160615.0       142
20170320.0       105
20161214.0       101
20170220.0        86
20150513.0        73
20170619.0        40
20170614.0        21
20171121.0        10
20161013.0   

In [13]:
teste['DATVEN'] = pd.to_datetime(teste['DATVEN'], format='%Y%m%d')
teste.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 414177 entries, 0 to 414176
Data columns (total 26 columns):
 #   Column  Non-Null Count   Dtype         
---  ------  --------------   -----         
 0   TIPREG  414177 non-null  int64         
 1   DATA-P  414177 non-null  datetime64[ns]
 2   CODBDI  414177 non-null  float64       
 3   CODNEG  414177 non-null  object        
 4   TPMERC  414177 non-null  int64         
 5   NOMRES  414177 non-null  object        
 6   ESPECI  414177 non-null  object        
 7   PRAZOT  249163 non-null  float64       
 8   MODREF  414177 non-null  object        
 9   PREABE  414177 non-null  float64       
 10  PREMAX  414177 non-null  float64       
 11  PREMIN  414177 non-null  float64       
 12  PREMED  414177 non-null  float64       
 13  PREULT  414177 non-null  float64       
 14  PREOFC  414177 non-null  float64       
 15  PREOFV  414177 non-null  float64       
 16  TOTNEG  414177 non-null  float64       
 17  QUATOT  414177 non-null  floa

In [21]:
teste['DATVEN'].value_counts()

2261-12-31    192692
2015-12-21     23477
2015-09-21     18099
2015-04-20     17599
2015-10-19     17250
2015-11-16     15836
2015-05-18     15291
2015-07-20     15144
2015-06-15     14557
2015-08-17     14537
2016-01-18     14332
2015-03-16     13961
2015-02-09      8243
2016-02-15      7151
2015-01-19      3948
2016-03-21      2859
2015-12-16      2671
2016-04-18      2145
2015-10-14      1722
2016-02-17      1652
2015-08-12      1251
2015-06-17      1231
2016-04-13      1082
2015-04-15       849
2016-12-19       630
2016-06-20       585
2016-05-16       580
2016-09-19       557
2016-10-17       545
2016-07-18       497
2015-02-18       456
2016-08-15       406
2016-11-21       333
2015-09-16       311
2015-11-18       293
2017-01-16       258
2016-08-17       214
2015-07-15       174
2016-01-13       156
2016-06-15       142
2017-03-20       105
2016-12-14       101
2017-02-20        86
2015-05-13        73
2017-06-19        40
2017-06-14        21
2017-11-21        10
2016-10-13   

### 8) Quais outros valores devem ser tratados? Examinar tabela B3.

        2   CODBDI  414177 non-null  float64 => integer - tabela

        7   PRAZOT  249163 non-null  float64 => integer

        16  TOTNEG  414177 non-null  float64 => integer

        17  QUATOT  414177 non-null  float64 => integer

        18  VOLTOT  414177 non-null  float64 => integer

        20  INDOPC  414177 non-null  float64 => ver tabela

        21  DATVEN  414177 non-null  float64 => data AAAAMMDD

        22  FATCOT  414177 non-null  float64 => integer

        25  DISMES  414177 non-null  float64 => integer
        
Irei fazer algumas suposições sobre os formatos abaixo, sem me extender sobre isso. No momento de uso dos dados deve-se verificar a adequação.

In [15]:
teste['CODBDI'] = teste['CODBDI'].astype('int')
teste['INDOPC'] = teste['INDOPC'].astype('int')
teste['FATCOT'] = teste['FATCOT'].astype('int')
teste['DISMES'] = teste['DISMES'].astype('int')
teste.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 414177 entries, 0 to 414176
Data columns (total 26 columns):
 #   Column  Non-Null Count   Dtype         
---  ------  --------------   -----         
 0   TIPREG  414177 non-null  int64         
 1   DATA-P  414177 non-null  datetime64[ns]
 2   CODBDI  414177 non-null  int64         
 3   CODNEG  414177 non-null  object        
 4   TPMERC  414177 non-null  int64         
 5   NOMRES  414177 non-null  object        
 6   ESPECI  414177 non-null  object        
 7   PRAZOT  249163 non-null  float64       
 8   MODREF  414177 non-null  object        
 9   PREABE  414177 non-null  float64       
 10  PREMAX  414177 non-null  float64       
 11  PREMIN  414177 non-null  float64       
 12  PREMED  414177 non-null  float64       
 13  PREULT  414177 non-null  float64       
 14  PREOFC  414177 non-null  float64       
 15  PREOFV  414177 non-null  float64       
 16  TOTNEG  414177 non-null  float64       
 17  QUATOT  414177 non-null  floa

8a) Não sei qual a diferença entre um zero e um NaN neste campo `'PRAZOT'`, mas imagino que haja já que há esta diferença.  
Encontrei a sugestão de substituir por outro valor, como `-1` por exemplo, e farei isso:

In [16]:
teste['PRAZOT'] = teste['PRAZOT'].fillna(-1)
teste['PRAZOT'] = teste['PRAZOT'].astype('int')
teste.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 414177 entries, 0 to 414176
Data columns (total 26 columns):
 #   Column  Non-Null Count   Dtype         
---  ------  --------------   -----         
 0   TIPREG  414177 non-null  int64         
 1   DATA-P  414177 non-null  datetime64[ns]
 2   CODBDI  414177 non-null  int64         
 3   CODNEG  414177 non-null  object        
 4   TPMERC  414177 non-null  int64         
 5   NOMRES  414177 non-null  object        
 6   ESPECI  414177 non-null  object        
 7   PRAZOT  414177 non-null  int64         
 8   MODREF  414177 non-null  object        
 9   PREABE  414177 non-null  float64       
 10  PREMAX  414177 non-null  float64       
 11  PREMIN  414177 non-null  float64       
 12  PREMED  414177 non-null  float64       
 13  PREULT  414177 non-null  float64       
 14  PREOFC  414177 non-null  float64       
 15  PREOFV  414177 non-null  float64       
 16  TOTNEG  414177 non-null  float64       
 17  QUATOT  414177 non-null  floa

8b) Assumindo que os campos abaixo também são *integer*:

In [17]:
teste['TOTNEG'] = teste['TOTNEG'].astype('int')
teste['QUATOT'] = teste['QUATOT'].astype('int')
teste['VOLTOT'] = teste['VOLTOT'].astype('int')
teste.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 414177 entries, 0 to 414176
Data columns (total 26 columns):
 #   Column  Non-Null Count   Dtype         
---  ------  --------------   -----         
 0   TIPREG  414177 non-null  int64         
 1   DATA-P  414177 non-null  datetime64[ns]
 2   CODBDI  414177 non-null  int64         
 3   CODNEG  414177 non-null  object        
 4   TPMERC  414177 non-null  int64         
 5   NOMRES  414177 non-null  object        
 6   ESPECI  414177 non-null  object        
 7   PRAZOT  414177 non-null  int64         
 8   MODREF  414177 non-null  object        
 9   PREABE  414177 non-null  float64       
 10  PREMAX  414177 non-null  float64       
 11  PREMIN  414177 non-null  float64       
 12  PREMED  414177 non-null  float64       
 13  PREULT  414177 non-null  float64       
 14  PREOFC  414177 non-null  float64       
 15  PREOFV  414177 non-null  float64       
 16  TOTNEG  414177 non-null  int64         
 17  QUATOT  414177 non-null  int6

#### Agora os únicos valores *float* são os preços.
Poderia-se fazer outras análises:

    TIPREG é um valor fixo, pode ser eliminado?
    Valores com tabelas (CODBDI,TPMERC,ESPECI,MODREF,INDOPC,FATCOT) são categóricos? 

In [18]:
# por exemplo
teste['INDOPC'].value_counts()
# deu 0 para tudo. Qual o significado? Nenhum contrato atual tem correção de preços?

0    414177
Name: INDOPC, dtype: int64

Tabela para `INDOPC`

        1 US$  CORREÇÃO PELA TAXA DO DÓLAR
        2 TJLP CORREÇÃO PELA TJLP
        8 IGPM CORREÇÃO PELO IGP-M - OPÇÕES PROTEGIDAS
        9 URV  CORREÇÃO PELA URV

## 9) Conclusão
Devido ao grande tamanho dos arquivos, seria recomendável armazená-los de outra maneira (por exemplo, em um banco de dados SQL).
Também é recomendável que os formatos de dados das colunas sejam verificados para que atendam ao necessário do trabalho.